### PR TITLE
[Op. Arithm.] Improve performance #2

### PR DESCRIPTION
**Pull request by AlbertMitjans**
Move expansion of matrix to the end. This way we avoid computing the sum/prod of very big matrices when `wire_order` contains extra wire values. 

This introduces amazing speed-ups when computing the matrix of a `Sum`/`Prod` with summands/factors that have different wire values.

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
Comparison of the time taken to compute the expectation value of the sum of 20 tensor products of 6 Pauli operators acting on 9 different wires:

- Master branch:
image


- This branch:
image

**Comment by AlbertMitjans** [comment]
Comparison of the time taken to compute the matrix of the sum of 20 tensor products of 6 Pauli operators acting on 9 different wires:

- Master branch:
image

- This branch:
image

**Review by Jaybsoni — COMMENTED** [review]
Once PR 1 in this series is merged in there won't be much new code here. I think coverage is still good with the tests we have, so not sure how you could even test the new lines of code. Honestly looks good to me. 

I will give it another look when PR 1 is in, but nothing blocking except for a changelog entry 😄

File: pennylane/ops/op_math/prod.py [hunk 1]
```diff
@@ -276,32 +286,47 @@ def diagonalizing_gates(self):
...
 
     def matrix(self, wire_order=None):
         """Representation of the operator as a matrix in the computational basis."""
-        wire_order = wire_order or self.wires
         mats = (
-            math.expand_matrix(qml.matrix(op), op.wires, wire_order=wire_order)
+            math.expand_matrix(qml.matrix(op), op.wires, wire_order=self.wires)
             if isinstance(op, qml.Hamiltonian)
-            else math.expand_matrix(op.matrix(), op.wires, wire_order=wire_order)
+            else math.expand_matrix(op.matrix(), op.wires, wire_order=self.wires)
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:320]:
```suggestion
            else op.matrix(wire_order=self.wires)
```

File: pennylane/ops/op_math/prod.py [hunk 2]
```diff
@@ -276,32 +286,47 @@ def diagonalizing_gates(self):
...
+        ]
+
+        return qml.math.prod(eigvals, axis=0)
 
     def matrix(self, wire_order=None):
         """Representation of the operator as a matrix in the computational basis."""
-        wire_order = wire_order or self.wires
         mats = (
-            math.expand_matrix(qml.matrix(op), op.wires, wire_order=wire_order)
+            math.expand_matrix(qml.matrix(op), op.wires, wire_order=self.wires)
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:318]:
```suggestion
            qml.matrix(op, wire_order=self.wires)
```

File: pennylane/ops/op_math/prod.py [hunk 3]
```diff
@@ -314,15 +314,19 @@ def eigvals(self):
...
             for op in self.factors
         )
 
-        return reduce(math.dot, mats)
+        reduced_mat = reduce(math.dot, mats)
+
+        if wire_order is not None:
+            reduced_mat = math.expand_matrix(reduced_mat, self.wires, wire_order=wire_order)
+
+        return reduced_mat
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:329]:
```suggestion
        return math.expand_matrix(reduced_mat, self.wires, wire_order=wire_order)
```

`math.expand_matrix` now performs a `wire_order is not None` check inside.

File: pennylane/ops/op_math/prod.py [hunk 4]
```diff
@@ -364,9 +368,11 @@ def _label(factor, decimals, base_label, cache):
...
     def sparse_matrix(self, wire_order=None):
         """Compute the sparse matrix representation of the Prod op in csr representation."""
-        wire_order = wire_order or self.wires
-        mats = (op.sparse_matrix(wire_order=wire_order) for op in self.factors)
-        return reduce(math.dot, mats)
+        mats = (op.sparse_matrix(wire_order=self.wires) for op in self.factors)
+        reduced_mat = reduce(math.dot, mats)
+        if wire_order is not None:
+            reduced_mat = math.expand_matrix(reduced_mat, self.wires, wire_order=wire_order)
+        return reduced_mat
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:375]:
```suggestion
        return math.expand_matrix(reduced_mat, self.wires, wire_order=wire_order)
```

File: pennylane/ops/op_math/sum.py [hunk 5]
```diff
@@ -314,18 +314,20 @@ def matrix(self, wire_order=None):
...
 
-        if wire_order is None:
-            wire_order = self.wires
+        reduced_mat = _sum(matrix_gen(self.summands))
 
-        return _sum(matrix_gen(self.summands, wire_order))
+        if wire_order is not None:
+            reduced_mat = math.expand_matrix(reduced_mat, self.wires, wire_order=wire_order)
+
+        return reduced_mat
```
Comment (albi3ro) [inline pennylane/ops/op_math/sum.py:330]:
```suggestion
        return math.expand_matrix(reduced_mat, self.wires, wire_order=wire_order)
```

File: pennylane/ops/op_math/sum.py [hunk 6]
```diff
@@ -366,9 +368,11 @@ def _label(factor, decimals, base_label, cache):
...
     def sparse_matrix(self, wire_order=None):
         """Compute the sparse matrix representation of the Sum op in csr representation."""
-        wire_order = wire_order or self.wires
-        mats_gen = (op.sparse_matrix(wire_order=wire_order) for op in self.summands)
-        return reduce(math.add, mats_gen)
+        mats_gen = (op.sparse_matrix(wire_order=self.wires) for op in self.summands)
+        reduced_matrix = reduce(math.add, mats_gen)
+        if wire_order is not None:
+            reduced_matrix = math.expand_matrix(reduced_matrix, self.wires, wire_order=wire_order)
+        return reduced_matrix
```
Comment (albi3ro) [inline pennylane/ops/op_math/sum.py:375]:
```suggestion
        return math.expand_matrix(reduced_matrix, self.wires, wire_order=wire_order)
```

**Review by albi3ro — APPROVED** [review]
Fascinating how such little changes can make such a huge difference.

Removing the `wire_order is not None` checks are an optional style improvement.

**Pull request merged**